rfcs/text/refactoring-date-component.md

```diff
@@ -0,0 +1,407 @@
+    displaySelectControls,
+    pickerProps
+  }) => {
+    const [lastValidValue, setLastValidValue] = useState(inputValue);
+    const { locale } = useContext(LocaleContext);
+    const { localize, options, formats } = localeMap[locale()];
+    const format = formats.javascript();
+    const { weekStartsOn } = options;
+
+    if (!isOpen) return null;
+
+    const renderCaptionElement = () => {
+      const { minDate, maxDate } = pickerProps;
+      return (
+        <YearMonthForm
+          date={selectedDate}
+          months={months}
+          onChange={setSelectedDate}
+          minDate={minDate}
+          maxDate={maxDate}
+        />
+      );
+    };
+
+    const captionElement = displaySelectControls ? renderCaptionElement() : undefined;
+
+    useEffect(() => {
+      // check if inputValue parses to valid date and update lastValidValue string if true
+    }, [
+      format,
+      inputValue,
+      lastValidValue,
+    ]);
+
+    return (
+      <Popover
+        ...
+      >
+      <StyledDatePicker>
+        <DayPicker
+          firstDayOfWeek={weekStartsOn}
+          onDayClick={onDayClick}
+          selectedDays={selectedDate}
+          month={parseToDate(lastValidValue, format)}
+          weekdayElement={renderWeekdayElement}
+          months={months}
+          captionElement={captionElement}
+          {...pickerProps}
+        />
+      </StyledDatePicker>
+    </Popover>
+    );
+  };
+```
+
+# Drawbacks
+
+Carrying out this proposal will mean introducing several breaking changes. Writing the new component as a functional component will mean that any consuming project that has extended the previous class-based component will now need to update to use composition instead. By moving the `Weekday` and `Navbar` components to an `__internal__` directory we will also be making a breaking change, but this will be for the better in the long run as these components should not be directly imported by consuming projects and we can then make changes to their interfaces without needing to have them be breaking. The `Date` component is used extensively by our consumers so these breaking changes are likely to have widespread impact. Although it is certainly arguable that the current implementation has enough issues with it that this too is having significant impact.
+
+The new proposed component will be tightly coupled to `date-fns` meaning if a better alternative becomes available, we will have a harder job migrating to it. However, writing our own internal utils that use the underlying `date-fns` interface should mitigate some of this work. Furthermore, if any issues arise with it we are reliant on support from the maintainers of `date-fns`, although it is widely adopted and maintained at present. The library uses different [Unicode patterns](https://github.com/date-fns/date-fns/blob/master/docs/unicodeTokens.md) for date formats so we will need to ensure these differences are communicated to implementation teams.
```

Comment:
```suggestion
The new proposed component will be tightly coupled to `date-fns` meaning if a better alternative becomes available, we will have a harder job migrating to it. However, writing our own internal utils that use the underlying `date-fns` interface should mitigate some of this work. Furthermore, if any issues arise with it we are reliant on support from the maintainers of `date-fns`, although it is widely adopted and maintained at present.
```

We don't need to communicate this as we're only using the `I18nProvider#locale()`